kubernetes/core/src/main/kotlin/com/google/kubernetes/tools/kubectl/KubectlExecutorService.kt

```diff
@@ -0,0 +1,162 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.kubernetes.tools.kubectl
+
+import com.intellij.execution.ExecutionException
+import com.google.cloud.tools.intellij.analytics.UsageTrackerService
+import com.intellij.openapi.components.ServiceManager
+import java.io.BufferedReader
+import java.io.InputStreamReader
+import java.util.concurrent.TimeUnit
+
+const val KUBECTL_SUCCESS = "kubectl.run"
+const val KUBECTL_FAIL = "kubectl.fail"
+const val METADATA_ERROR_MESSAGE_KEY = "error.message"
+
+class KubectlExecutorService {
+    companion object {
+        val instance
+            get() = ServiceManager.getService(KubectlExecutorService::class.java)!!
+    }
+
+    var kubectlExecutablePath: String = "kubectl"
+
+    /**
+     * run kubectl version to make sure that kubectl is available, if there's no response after
+     *  a couple of seconds then it is not available.
+     */
+    fun isKubectlAvailable(): Boolean =
+        try {
+            val settings = KubectlExecutorSettings(KubectlExecutorSettings.ExecutionMode.VERSION)
+            val process = executeKubectl(settings).process
+            process.waitFor(1, TimeUnit.SECONDS)
+            process.exitValue() == 0
+        } catch (e: Exception) {
+            false
+        }
+
+    /**
+     * Parse through command line arguments and try to execute the command
+     */
+    fun executeKubectl(settings: KubectlExecutorSettings): KubectlProcess {
+
+        val commandList = mutableListOf<String>()
+
+        with(commandList) {
+            add(kubectlExecutablePath)
+            add(settings.executionMode.modeFlag)
+            settings.executionFlags.forEach {
+                add(it)
+            }
+        }
+
+        try {
+            val processBuilder = ProcessBuilder()
+            val startedProcess = processBuilder.command(commandList).start()
+            val kubectlProcess = KubectlProcess(
+                    startedProcess,
+                    commandList.joinToString(" ")
+            )
+            return kubectlProcess
+        } catch (e: Exception) {
+            val kubectlFailEventName = KUBECTL_FAIL
+
+            UsageTrackerService.getInstance().trackEvent(kubectlFailEventName)
+                    .addMetadata(
+                            METADATA_ERROR_MESSAGE_KEY, e.javaClass.name
+                    ).ping()
+            throw e
+        }
+    }
+
+    /**
+     * Check if kubectl is available, Check if the flags supplied are valid. Create and start the
+     * command line process passed in
+     */
+    fun processOutputToString(executingProcess: Process): String {
```

Comment:
If this method is only used internally and for tests, can we mark it `@VisibleForTesting` since it's public only for unit tests, and move it below the next one, which is the public API call?

Reply:
I'll do that. It's not only used for the tests though, it's used by startProcess as well.

Reply:
startProcess is the only api call though. Should all the other functions also be marked @VisibleForTesting ?

Reply:
generally you should only mark functions `@VisibleForTesting` if the are exposed (made non-private) _only for testing purposes_. If the other functions can be made private, then they should be. If not, then only mark them `@VisibleForTesting` if the only reason for not doing so is for unit testing.

Reply:
+1 to Etan, great summary.